tuxedo-core/src/verifier/htlc.rs

```diff
@@ -0,0 +1,380 @@
+///
+/// Decision: I will add block number to the signature. And remain open to adding more blockchain-level
+/// fundamental things. Especially if they are available in bitcoin script.
+///
+/// Regarding the verifier constraint checker separation, perhaps the right line to be drawn is
+/// that verifiers are useful in a lot of places, but perhaps not expressive enough in others.
+/// When they are not expressive enough, just use `UpForGrabs` and rely on the constraint checker,
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct TimeLock {
+    pub unlock_block_height: u32,
+}
+
+impl Verifier for TimeLock {
+    type Redeemer = ();
+    fn verify(&self, _: &[u8], block_height: u32, _: &()) -> bool {
+        block_height >= self.unlock_block_height
+    }
+}
+
+/// Allows UTXOs to be spent when a preimage to a recorded hash is provided.
+/// This could be used as a puzzle (although a partial preimage search would be better)
+/// or a means of sharing a password, or as part of a simple atomic swapping protocol.
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct BlakeTwoHashLock {
+    pub hash_lock: H256,
+}
+
+impl BlakeTwoHashLock {
+    pub fn new_from_secret<T: Encode>(secret: T) -> Self {
+        Self {
+            hash_lock: BlakeTwo256::hash(&secret.encode()),
+        }
+    }
+}
+
+impl Verifier for BlakeTwoHashLock {
+    type Redeemer = Vec<u8>;
+    fn verify(&self, _: &[u8], _: u32, secret: &Self::Redeemer) -> bool {
+        BlakeTwo256::hash(secret) == self.hash_lock
+    }
+}
+
+/// Allows a UTXO to be spent, and therefore acknowledged by an intended recipient by revealing
+/// a hash preimage. After an initial claim period elapses on chain, the UTXO can also be spent
+/// by the refunder. In practice, the refunder is often the same address initially funded the HTLC.
+///
+/// The receiver and refunder are specified as a simple public keys for simplicity. It would be
+/// interesting to use public key hash, or better yet, simply abstract this over some opaque
+/// inner verifier for maximum composability.
+///
+/// @lederstrumpf when the time expires, is the primary "happy" path supposed to remain open?
+/// Or is it that if the swap hasn't started on time, the refund is the only option.
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct HashTimeLockContract {
+    /// The hash whose preimage must be revealed (along with the recipient's signature) to spend the UTXO.
+    pub hash_lock: H256,
+    /// The pubkey that is intended to receive and acknowledge receipt of the funds.
+    pub recipient_pubkey: Public,
+    /// The time (as a block height) when the refund path opens up.
+    pub claim_period_end: u32,
+    /// The address who can spend the coins without revealing the preimage after the claim period has ended.
+    pub refunder_pubkey: Public,
+}
+
+///
```

Comment:
I think you forgot some comments here :)

Reply:
Fixed in e9fb8d85d2e6